src/linear_algebra/free_module/ideal_quotient.lean

```diff
@@ -98,3 +99,29 @@ let b := module.free.choose_basis ℤ S,
 in by haveI : ∀ i, ne_zero (a i).nat_abs :=
     (λ i, ⟨int.nat_abs_ne_zero_of_ne_zero (ideal.smith_coeffs_ne_zero b I hI i)⟩); classical;
   exact fintype.of_equiv (Π i, zmod (a i).nat_abs) e.symm
+
+variables (F : Type*) [comm_ring F] [algebra F R] [algebra F S] [is_scalar_tower F R S]
+  (b : basis ι R S) {I : ideal S} (hI : I ≠ ⊥)
+
+/-- Decompose `S⧸I` as a direct sum of cyclic `R`-modules
+  (quotients by the ideals generated by Smith coefficients of `I`). -/
+noncomputable def ideal.quotient_equiv_direct_sum :
+  (S ⧸ I) ≃ₗ[F] ⨁ i, R ⧸ ideal.span ({I.smith_coeffs b hI i} : set R) :=
+begin
+  apply ((I.quotient_equiv_pi_span b _).restrict_scalars F).trans
+    (direct_sum.linear_equiv_fun_on_fintype _ _ _).symm,
+  exact linear_map.is_scalar_tower.compatible_smul
+  -- why doesn't it automatically apply?
+  -- even after `change linear_map.compatible_smul _ (Π i, R ⧸ span _) F R`
+end
+
+lemma ideal.finrank_quotient_eq_sum [nontrivial F]
+  [∀ i, module.free F (R ⧸ ideal.span ({I.smith_coeffs b hI i} : set R))]
+  [∀ i, module.finite F (R ⧸ ideal.span ({I.smith_coeffs b hI i} : set R))] :
+  finite_dimensional.finrank F (S ⧸ I)
+    = ∑ i, finite_dimensional.finrank F (R ⧸ ideal.span ({I.smith_coeffs b hI i} : set R)) :=
+begin
+  rw [linear_equiv.finrank_eq $ ideal.quotient_equiv_direct_sum F b hI,
+      finite_dimensional.finrank_direct_sum]
+  -- slow, and dot notation doesn't work
+end
```

Comment:
This lemma might look unnecessary but I just tested and it can't be inlined in [`finrank_quotient_span_eq_nat_degree_norm`](https://github.com/leanprover-community/mathlib/blob/71180dbf8a7cbcdd87b99d3aea18a61bb91aace5/src/algebraic_geometry/elliptic_curve/auxiliary.lean#L120) in #18000.